### PR TITLE
Cache closed comment threads for longer.

### PR DESCRIPTION
**Pull request by gklopper**
## What does this change?
If comment threads are closed, they are cached for slightly more than an hour. If they are open they remain cached for 60 seconds.

## What is the value of this and can you measure success?
I have been parsing logs of cache misses and this class of endpoints is one of the biggest in there.

We will be able to measure lower load on the discussion backend. Depending on how much lower it is, this will also equate to cpu and bandwidth savings.

## Does this affect other platforms - Amp, Apps, etc?
No

### CODE REVIEW

**Comment by gklopper** [comment]
cc @johnduffell

File: static/src/javascripts/projects/common/modules/discussion/loader.js [hunk 1]
```diff
@@ -69,7 +69,7 @@ Loader.prototype.initTopComments = function() {
         this.gotoComment(commentId);
     });
 
-    return fetchJson('/discussion/top-comments/' + this.getDiscussionId() + '.json', {
+    return fetchJson('/discussion/top-comments/' + this.getDiscussionId() + '.json?closed=' + config.page.commentable, {
```
Comment (gklopper) [inline static/src/javascripts/projects/common/modules/discussion/loader.js]: There is no whitelist on api.nextgen.guardianapps.co.uk. The whitelist on www.theguardian.com is mainly to deal with campaign codes (which don't apply on this endpoint).

Comment (dominickendrick) [inline static/src/javascripts/projects/common/modules/discussion/loader.js]: 🎉

File: discussion/app/controllers/CommentsController.scala [hunk 2]
```diff
@@ -161,4 +162,11 @@ class CommentsController(csrfConfig: CSRFConfig, val discussionApi: DiscussionAp
       }
     } recover toResult
   }
+
+  // caches "closed" comment threads for an hour.
+  // if the thread is switched on again the url changes and it cache busts itself.
+  private def cacheTime(request: RequestHeader, defaultCacheSeconds: Int) = {
+    val commentsClosed = request.getParameter("closed").contains("true")
+    if (commentsClosed && LongCacheCommentsSwitch.isSwitchedOn) 3800 else defaultCacheSeconds
```
Comment (johnduffell) [inline discussion/app/controllers/CommentsController.scala]: ideally keep all integer cache seconds inside Cached.scala with the others, as a val with a helpful name

**Comment by johnduffell** [comment]
good spot! go for it 👍 🚀

File: discussion/app/controllers/CommentsController.scala [hunk 3]
```diff
@@ -161,4 +162,11 @@ class CommentsController(csrfConfig: CSRFConfig, val discussionApi: DiscussionAp
       }
     } recover toResult
   }
+
+  // caches "closed" comment threads for an hour.
+  // if the thread is switched on again the url changes and it cache busts itself.
+  private def cacheTime(request: RequestHeader) = {
+    val commentsClosed = request.getParameter("closed").contains("true")
+    if (commentsClosed && LongCacheCommentsSwitch.isSwitchedOn) CacheTime.DiskCached else CacheTime.Default
```
Comment (johnduffell) [inline discussion/app/controllers/CommentsController.scala]: as opposed to `CacheTime.ClosedComments else CacheTime.OpenComments`?

File: discussion/app/controllers/CommentsController.scala [hunk 4]
```diff
@@ -161,4 +162,11 @@ class CommentsController(csrfConfig: CSRFConfig, val discussionApi: DiscussionAp
       }
     } recover toResult
   }
+
+  // caches "closed" comment threads for an hour.
+  // if the thread is switched on again the url changes and it cache busts itself.
+  private def cacheTime(request: RequestHeader) = {
+    val commentsClosed = request.getParameter("closed").contains("true")
+    if (commentsClosed && LongCacheCommentsSwitch.isSwitchedOn) CacheTime(3800) else CacheTime(60)
```
Comment (johnduffell) [inline discussion/app/controllers/CommentsController.scala]: 😆

**Pull request merged**